File: public/tictactoe.js
'use strict'

import { createApp, reactive, shallowReactive } from 'https://cdn.jsdelivr.net/npm/vue@3/dist/vue.esm-browser.js'
import Module from './modules/demos.js'

(async () => {
  const wasmModule = await Module()

  let gridWorld = new wasmModule.GridWorld(3, 3)
  let board = new wasmModule.TicTacToe()
  let bot = new wasmModule.Bot(board, 2)

  const app = createApp({
    data: () => ({
      gridWorld: shallowReactive(gridWorld),
      board: shallowReactive(board),
      bot: shallowReactive(bot),
    }),
    computed: {
      getWidth() {
        return this.gridWorld.width_readonly
      },
      getHeight() {
        return this.gridWorld.height_readonly
      },
      locations() {
        const keys = this.gridWorld.locations().keys()
        const result = []
        for (let i = 0; i < keys.size(); i++) {
          const location = keys.get(i)
          result.push(location)
        }
        return result
      },
    },
    methods: {
      get(location) {
        return this.board.get(location.x, location.y)
      },
      set(location) {
        let who = this.board.has_winner();
        if (who != 0 || this.board.is_complete()) {
          return;
        }

        let where = this.board.get(location.x, location.y)
        if (where != 0) {
          return;
        }

        this.board.set(location.x, location.y, 1)
        this.bot.best_move()
        this.$forceUpdate()
      },
      reset() {
        this.board.reset()
        this.$forceUpdate()
      },

    },
  })

  app.provide("demos", wasmModule)
  app.mount(document.querySelector("#diagram1"))
})()

File: demos/tictactoe/tictactoe.cpp
#include "tictactoe.h"

namespace TicTacToe
{

Board::Board()
{
    Reset();
}

void Board::Reset()
{
    for (auto& row : board)
    {
        row.fill(EMPTY);
    }
}

int Board::HasWinner() const
{
    for (int i = 0; i < 3; ++i)
    {
        if (board[i][0] != EMPTY && board[i][0] == board[i][1] && board[i][0] == board[i][2])
        {
            return board[i][0];
        }
        if (board[0][i] != EMPTY && board[0][i] == board[1][i] && board[0][i] == board[2][i])
        {
            return board[0][i];
        }
    }
    if (board[1][1] != EMPTY &&
        ((board[0][0] == board[1][1] && board[2][2] == board[1][1]) ||
         (board[0][2] == board[1][1] && board[2][0] == board[1][1])))
    {
        return board[1][1];
    }
    return EMPTY;
}

bool Board::IsComplete() const
{
    for (const auto& row : board)
    {
        for (int cell : row)
        {
            if (cell == EMPTY)
                return false;
        }
    }
    return true;
}

int Board::Get(int x, int y) const
{
    return board[x][y];
}

void Board::Set(int x, int y, int value)
{
    board[x][y] = value;
}

} // namespace TicTacToe